File: ios/BitcoinAddressGenerator.m
//
//  BitcoinAddressGenerator.m
//  bitkit
//
//  Created by Corey Phillips on 4/2/24.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(BitcoinAddressGenerator, NSObject)

RCT_EXTERN_METHOD(sayHello:(RCTResponseSenderBlock)callback)

RCT_EXTERN_METHOD(getAddress:(NSString *)mnemonic
                  derivationPath:(NSString *)derivationPath
                  networkType:(NSString *)networkType
                  callback:(RCTResponseSenderBlock)callback)

RCT_EXTERN_METHOD(getPrivateKey:(NSString *)mnemonic
                  derivationPath:(NSString *)derivationPath
                  networkType:(NSString *)networkType
                  bip39Passphrase:(NSString *)bip39Passphrase
                  callback:(RCTResponseSenderBlock)callback)

RCT_EXTERN_METHOD(getScriptHash:(NSString *)address
                  network:(NSString *)network
                  callback:(RCTResponseSenderBlock)callback)

@end

File: ios/BitcoinAddressGenerator.swift
//
//  BitcoinAddressGenerator.swift
//  bitkit
//
//  Created by Corey Phillips on 4/2/24.
//

import Foundation

@objc(BitcoinAddressGenerator)
class BitcoinAddressGenerator: NSObject {

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    @objc func sayHello(_ callback: RCTResponseSenderBlock) {
        callback([sayHi()])
    }

    @objc func getAddress(_ mnemonic: String, derivationPath: String, networkType: String, callback: RCTResponseSenderBlock) {
      let result = bitkit.getAddress(mnemonic: mnemonic, derivationPath: derivationPath, networkType: networkType)
        callback([result])
    }
  
    @objc func getPrivateKey(_ mnemonic: String, derivationPath: String, networkType: String, bip39Passphrase: String, callback: RCTResponseSenderBlock) {
        let result = bitkit.getPrivateKey(mnemonic: mnemonic, derivationPath: derivationPath, networkType: networkType, bip39Passphrase: bip39Passphrase)
        callback([result])
    }

    @objc func getScriptHash(_ address: String, network: String, callback: RCTResponseSenderBlock) {
        let result = bitkit.getScriptHash(address: address, network: network)
        callback([result])
    }
}
